src/custom/hooks/useTokensLazy.ts

```diff
@@ -0,0 +1,92 @@
+import { useCallback } from 'react'
+
+import { Token } from '@uniswap/sdk-core'
+
+import { useActiveWeb3React } from '@src/hooks/web3'
+import { getBytes32TokenContract, getTokenContract, useMulticall2Contract } from 'hooks/useContract'
+import { CallParams, getMultipleCallsResults } from 'state/multicall/utils'
+import { parseStringOrBytes32 } from '@src/hooks/Tokens'
+import { useAddUserToken } from '@src/state/user/hooks'
+
+/**
+ * Hook that exposes a function for lazy (non-hook) fetching tokens info async
+ *
+ * Similar to useToken, except it fetches multiple tokens at once, and does it after component hook phase
+ */
+export function useTokensLazy() {
+  const { library, account, chainId } = useActiveWeb3React()
+  const multicall2Contract = useMulticall2Contract()
+  const addUserToken = useAddUserToken()
+
+  return useCallback(
+    async (addresses: string[], withSignerIfPossible?: boolean): Promise<Record<string, Token | null> | null> => {
+      if (!library || !account || !chainId || !multicall2Contract) {
+        console.warn(`useTokensLazy::not initialized`)
+        return null
+      }
+
+      // TODO: maybe we'll need to bach these queries if searching for too many tokens at once?
+      const callsParams = addresses.reduce<CallParams[]>((acc, address) => {
+        const tokenContract = getTokenContract(address, withSignerIfPossible, library, account, chainId)
+        const tokenContractBytes32 = getBytes32TokenContract(address, withSignerIfPossible, library, account, chainId)
+
+        // For every address, create 5 queries
+        // Merge them into a single array to be used by the multicall contract
+        return acc.concat([
+          {
+            contract: tokenContract,
+            methodName: 'name',
+          },
+          {
+            contract: tokenContract,
+            methodName: 'symbol',
+          },
+          {
+            contract: tokenContract,
+            methodName: 'decimals',
+          },
+          {
+            contract: tokenContractBytes32,
+            methodName: 'name',
+          },
+          {
+            contract: tokenContractBytes32,
+            methodName: 'symbol',
+          },
+        ])
+      }, [])
+
+      // Single multicall to all tokens
+      const results = await getMultipleCallsResults({
```

Comment:
I have a TODO for that, and I'll check how the base code splits multicall into chuncks.
Still, I do not expect to have enough here to be a problem.

Was talking to @Uxio0 about that (since now they are using multicall on the Safe to fetch balances) and according to him, over 100 token updates in a single call works without a problem. (Right Uxio?)

Reply:
That's right, for some Safes we call `balanceOf` for more than 100 tokens in the same `Multicall`, but I would recommend to split the call if possible, as for example `Geth` nodes have an internal timeout of 5 seconds and sometimes it can take more than that (my assumption is that it depends of how busy the node is)

Reply:
Agree @anxolin, with an unkown node can be trickier